applications/web/src/components/tools/NewLine.svelte

```diff
@@ -8,46 +8,43 @@
 
   export let pointsById: IDictionary<SketchPoint>, sketchIndex: string, active: boolean, projectToPlane: ProjectToPlane
 
-  // $: pointsById, log("[props]", pointsById, sketchIndex, active, projectToPlane)
+  let previousPoint: PointLikeById | null = null
 
-  let previousPoint: PointLikeById | null
+  let stack: PointLikeById[] = []
 
-  $: if ($sketchTool !== "line") previousPoint = null
+  $: if ($sketchTool !== "line") clearStack()
 
-  function processPoint(point: PointLikeById | null) {
-    if (!previousPoint && point) {
-      // if there is no anchor point, set one
-      if (point.id) {
-        // nothing to do, the point exists!
-        log("[processPoint] nothing to do the point exists!")
-      } else {
-        log("[processPoint] oh cool, creating point!")
-        point.id = null
-      }
-    } else {
-      // there WAS an anchor point, so we should create a line
-
-      // if the center point doesn't exist, then we should create a point
-      if (previousPoint?.id === null) previousPoint.id = addPointToSketch(sketchIndex, previousPoint?.twoD!, false)
-
-      if (point?.id) {
-        // if the point exists, then we should create a line
-        if (previousPoint?.id && point.id) addLineToSketch(sketchIndex, +previousPoint.id, +point.id)
-        previousPoint = null
-        return
-      } else {
-        // if the point doesn't exist, then we should create a point and a line
-        point!.id = addPointToSketch(sketchIndex, point!.twoD!, false)
-        addLineToSketch(sketchIndex, +previousPoint!.id!, +point!.id!)
-      }
+  function pushToStack(point: PointLikeById) {
+    if (!point) return
+    if (!point.id) point.id = addPointToSketch(sketchIndex, point.twoD, false)
+    stack.push(point)
+  }
+
+  function processPoint(point: PointLikeById) {
+    pushToStack(point)
+
+    switch (stack.length) {
+      case 0: // nothing to do, the stack is empty
+        break
+      case 1: // can't create a line with only one point!
```

Comment:
wow that's so much clearer! great approach!

Reply:
Thanks! Just had to write something my brain could understand haha